### PR TITLE
fixes typing in pytorch_lightning/callbacks/stochastic_weight_avg.py

### PR DESCRIPTION
**Pull request by donlapark**
## What does this PR do?

Fixes typing in pytorch in `pytorch_lightning/callbacks/stochastic_weight_avg.py` from #13445.

Note:
* Line 109: there has been a known issue of `isinstance(..., Callable)` throwing an arg-type error (https://github.com/python/mypy/issues/6864).
* Line 127: If `self._swa_epoch_start` is a `float`, it is expected to be converted to `int` by a call to `on_fit_start()`. However, `mypy` does not see this and throws in a return-value error.
* Line 175: `torch.optim.swa_utils.SWALR` has not been type-annotated yet, so the type of the argument of `swa_lr` is automatically set to `float`.
* Line 246-258: `_BatchNorm.running_mean`, `_BatchNorm.running_var` and `_BatchNorm.num_batches_tracked` have types `Optional[Tensor]`. These attributes are initialized to be zero tensors via `self.reset_parameters()` but `mypy` does not see this.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/pytorch_lightning/callbacks/stochastic_weight_avg.py [hunk 1]
```diff
@@ -106,7 +106,7 @@ def __init__(
         if wrong_type or wrong_float or wrong_list:
             raise MisconfigurationException("The `swa_lrs` should a positive float, or a list of positive floats")
 
-        if avg_fn is not None and not isinstance(avg_fn, Callable):
+        if avg_fn is not None and not isinstance(avg_fn, Callable):  # type: ignore[arg-type]
```
Comment (otaj) [inline src/pytorch_lightning/callbacks/stochastic_weight_avg.py:109]: How about using a `callable` check? https://docs.python.org/3/library/functions.html#callable

File: src/pytorch_lightning/callbacks/stochastic_weight_avg.py [hunk 2]
```diff
@@ -118,27 +118,28 @@ def __init__(
...
-        self._model_contains_batch_norm = None
-        self._average_model = None
+        self._max_epochs: int
+        self._model_contains_batch_norm: bool
+        self._average_model = pl.LightningModule()
 
     @property
     def swa_start(self) -> int:
-        return max(self._swa_epoch_start - 1, 0)  # 0-based
+        return max(self._swa_epoch_start - 1, 0)  # type: ignore[return-value]
```
Comment (otaj) [inline src/pytorch_lightning/callbacks/stochastic_weight_avg.py:127]: Please do not remove comments

Comment (otaj) [reply src/pytorch_lightning/callbacks/stochastic_weight_avg.py:127]: And by the time it is used, it will already have the type properly resolved (no more float, but int only)

Comment (donlapark) [reply src/pytorch_lightning/callbacks/stochastic_weight_avg.py]: My bad. The comment has been added back.

**Review by Borda — APPROVED** [review]

File: src/pytorch_lightning/callbacks/stochastic_weight_avg.py (hunk 2 above)
Comment (akihironitta) [reply src/pytorch_lightning/callbacks/stochastic_weight_avg.py]: > And by the time it is used, it will already have the type properly resolved (no more float, but int only)

@otaj I thought so, too, but mypy seems to miss the value range check in `__init__`.

File: src/pytorch_lightning/callbacks/stochastic_weight_avg.py [hunk 3]
```diff
@@ -118,27 +118,28 @@ def __init__(
         self._annealing_strategy = annealing_strategy
         self._avg_fn = avg_fn or self.avg_fn
         self._device = device
-        self._model_contains_batch_norm = None
-        self._average_model = None
+        self._max_epochs: int
+        self._model_contains_batch_norm: bool
+        self._average_model = pl.LightningModule()
```
Comment (akihironitta) [inline src/pytorch_lightning/callbacks/stochastic_weight_avg.py:123]: Instead, I think we could do:
```suggestion
        self._average_model: pl.LightningModule
```

Comment (donlapark) [reply src/pytorch_lightning/callbacks/stochastic_weight_avg.py:123]: Maybe like this, just to be on the safe side?

```suggestion
        self._average_model: "pl.LightningModule"
```

File: src/pytorch_lightning/callbacks/stochastic_weight_avg.py (hunk 2 above)
Comment (otaj) [reply src/pytorch_lightning/callbacks/stochastic_weight_avg.py]: it doesn't "miss" anything, the variable is a `Union[int, float]` in `__init__`, it only gets forced to `int` in `on_fit_start` https://github.com/Lightning-AI/lightning/blob/26361adce107234a858de7bd3a163bef43832ffb/src/pytorch_lightning/callbacks/stochastic_weight_avg.py#L149.

But it is first used by `on_train_epoch_start`, so that should be resolved by then. But, mypy doesn't know (and cannot know) about this order

Comment (donlapark) [reply src/pytorch_lightning/callbacks/stochastic_weight_avg.py]: So what is the "right" way to resolve this? A simple fix would be put `# type: ignore` there or `assert isinstance(self._swa_epoch_start, int)`.

Comment (otaj) [reply src/pytorch_lightning/callbacks/stochastic_weight_avg.py]: I'd go with assert

Comment (otaj) [reply src/pytorch_lightning/callbacks/stochastic_weight_avg.py]: And check that the assert doesn't break any tests of course :)

**Review by kaushikb11 — APPROVED** [review]

File: src/pytorch_lightning/callbacks/stochastic_weight_avg.py [hunk 4]
```diff
@@ -229,35 +234,39 @@ def on_train_end(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule"):
...
-                module.running_mean, device=pl_module.device, dtype=module.running_mean.dtype
+                module.running_mean,  # type: ignore[arg-type]
+                device=pl_module.device,
+                dtype=module.running_mean.dtype,  # type: ignore[union-attr]
             )
             module.running_var = torch.ones_like(
-                module.running_var, device=pl_module.device, dtype=module.running_var.dtype
+                module.running_var,  # type: ignore[arg-type]
+                device=pl_module.device,
+                dtype=module.running_var.dtype,  # type: ignore[union-attr]
```
Comment (carmocca) [inline src/pytorch_lightning/callbacks/stochastic_weight_avg.py:255]: IMO these shouldn't have been ignored, but instead added assertions that `running_{mean,var}` are not None

File: src/pytorch_lightning/callbacks/stochastic_weight_avg.py [hunk 5]
```diff
@@ -229,35 +234,39 @@ def on_train_end(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule"):
...
             )
             self.momenta[module] = module.momentum
-            module.momentum = None
-            module.num_batches_tracked *= 0
+            module.momentum = None  # type: ignore[assignment]
+            module.num_batches_tracked *= 0  # type: ignore[assignment, operator]
```
Comment (carmocca) [inline src/pytorch_lightning/callbacks/stochastic_weight_avg.py:259]: Same comment here about asserting that it's not None

File: src/pytorch_lightning/callbacks/stochastic_weight_avg.py [hunk 6]
```diff
@@ -167,12 +169,15 @@ def on_train_epoch_start(self, trainer: "pl.Trainer", pl_module: "pl.LightningMo
             for lr, group in zip(self._swa_lrs, optimizer.param_groups):
                 group["initial_lr"] = lr
 
-            self._swa_scheduler = SWALR(
-                optimizer,
-                swa_lr=self._swa_lrs,
-                anneal_epochs=self._annealing_epochs,
-                anneal_strategy=self._annealing_strategy,
-                last_epoch=trainer.max_epochs if self._annealing_strategy == "cos" else -1,
+            self._swa_scheduler: _LRScheduler = cast(
```
Comment (carmocca) [inline src/pytorch_lightning/callbacks/stochastic_weight_avg.py:172]: The type annotation is redundant when you cast

```suggestion
            self._swa_scheduler = cast(
```

**Review by carmocca — COMMENTED** [review]
You can open a follow-up PR with the suggestions applied
